tests/flags/default_values.cc

```diff
@@ -0,0 +1,161 @@
+#include <array>
+
+#include "gtest/gtest.h"
+#include "stout/flags/flags.h"
+#include "tests/flags/test_default_values.pb.h"
+
+TEST(FlagsTest, FlagsWithDefaultValues) {
+  test::default_values::Flags flags;
+
+  auto parser = stout::flags::Parser::Builder(flags).Build();
+
+  std::array arguments = {
+      "/path/to/program",
+      "no-bam",
+  };
+
+  int argc = arguments.size();
+  const char** argv = arguments.data();
+
+  parser.Parse(&argc, &argv);
+
+  EXPECT_EQ(1, argc);
+  EXPECT_STREQ("/path/to/program", argv[0]);
+  EXPECT_EQ("'some default value'", flags.str1());
+  EXPECT_EQ("some default value", flags.str2());
+  EXPECT_EQ("\"some default value\"", flags.str3());
+  EXPECT_TRUE(flags.bar());
+  EXPECT_EQ(1994, flags.baz());
+  EXPECT_FALSE(flags.bam());
+}
+
+TEST(FlagsTest, PositionalArgsWithDefaultValues) {
+  test::default_values::PositionalArgs args;
+
+  auto parser = stout::flags::Parser::Builder(args).Build();
+
+  std::array arguments = {
+      "/path/to/program",
+      "13",
+  };
+
+  int argc = arguments.size();
+  const char** argv = arguments.data();
+
+  parser.Parse(&argc, &argv);
+
+  EXPECT_EQ(1, argc);
+  EXPECT_STREQ("/path/to/program", argv[0]);
+  EXPECT_EQ("13", args.arg1());
+  EXPECT_EQ("value2", args.arg2());
+  EXPECT_EQ("value3", args.arg3());
+}
+
+TEST(FlagsTest, MissingRequiredFlagWithDefaultValue) {
```

Comment:
Hmm, how does / should this combination work? (Why have a default value for a required flag, where the user has to provide a value?)

Reply:
It seems to me that a flag/arg that's both required and has a default value is a bit broken, since we'll never use the default value (since users must always override it when specifying the flag on the command line).

Is that understanding correct?

If so, it feels worth adding some validation to tell users that their flag/arg configuration (having one that's both required and default) is non-sensical (since that default value they provided will never be used), and instead we should make users pick a sensical configuration (probably just required with no default value).

Does that make sense?